[🐛 BUG] PrincipalDetails 객체를 RedisOauth2AuthorizationService에서 JSON 역직렬화를 할 수 없는 문제를 해결한다.

### 멈춰! 🖐🏻 잠시 체크 들어갑니다~

- [X] [다음 페이지](https://github.com/seeyouletter/seeyouletter-be/issues)에서 중복되지 않음을 확인했어요. 🙆🏻
- [X] 클라이언트 쪽 문제일 경우, [프론트엔드 레포지토리](https://github.com/seeyouletter/seeyouletter-fe)도 확인해보셨나요?
- [X] 사안을 볼 때 매우 위험해요. 빠른 수정이 필요해요! 😖

### 현재 버그는 어떤 주제인가요? (필수)

회원 인증 관련

### 🎬 문제 재현

1. 씨유레터에 씨유레터, 네이버, 카카오 중 어떤 방식으로든 로그인을 한다.
2. 인가 요청을 한다.
3. 인가 요청을 통해 받은 접근 토큰으로 엑세스 토큰을 요청한다.

API Response

```json
{
    "timestamp": "2023-02-25T04:53:36.371+00:00",
    "status": 500,
    "error": "Internal Server Error",
    "path": "/oauth2/token"
}
```

Stack Trace
```text
java.lang.IllegalArgumentException: The class with com.seeyouletter.api_member.auth.config.PrincipalDetails and name of com.seeyouletter.api_member.auth.config.PrincipalDetails is not in the allowlist. If you believe this class is safe to deserialize, please provide an explicit mapping using Jackson annotations or by providing a Mixin. If the serialization is only done by a trusted source, you can also enable default typing. See https://github.com/spring-projects/spring-security/issues/4370 for details (through reference chain: com.seeyouletter.api_member.config.RedisOauth2AuthorizationService$Authorization["attributes"])
	at com.seeyouletter.api_member.config.RedisOauth2AuthorizationService.deserialize(RedisOauth2AuthorizationService.java:494) ~[classes/:na]
	at com.seeyouletter.api_member.config.RedisOauth2AuthorizationService.findByToken(RedisOauth2AuthorizationService.java:139) ~[classes/:na]
	at org.springframework.security.oauth2.server.authorization.authentication.CodeVerifierAuthenticator.authenticate(CodeVerifierAuthenticator.java:81) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.security.oauth2.server.authorization.authentication.CodeVerifierAuthenticator.authenticateRequired(CodeVerifierAuthenticator.java:63) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.security.oauth2.server.authorization.authentication.PublicClientAuthenticationProvider.authenticate(PublicClientAuthenticationProvider.java:95) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:182) ~[spring-security-core-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.web.OAuth2ClientAuthenticationFilter.doFilterInternal(OAuth2ClientAuthenticationFilter.java:122) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.web.NimbusJwkSetEndpointFilter.doFilterInternal(NimbusJwkSetEndpointFilter.java:85) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.oidc.web.OidcProviderConfigurationEndpointFilter.doFilterInternal(OidcProviderConfigurationEndpointFilter.java:86) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.web.OAuth2AuthorizationEndpointFilter.doFilterInternal(OAuth2AuthorizationEndpointFilter.java:156) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.web.OAuth2AuthorizationServerMetadataEndpointFilter.doFilterInternal(OAuth2AuthorizationServerMetadataEndpointFilter.java:84) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:103) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:89) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:117) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.CorsFilter.doFilterInternal(CorsFilter.java:91) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:112) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:82) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.oauth2.server.authorization.config.annotation.web.configurers.AuthorizationServerContextFilter.doFilterInternal(AuthorizationServerContextFilter.java:61) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:55) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.session.DisableEncodeUrlFilter.doFilterInternal(DisableEncodeUrlFilter.java:42) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:221) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:186) ~[spring-security-web-5.7.5.jar:5.7.5]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267) ~[spring-web-5.3.24.jar:5.3.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.3.24.jar:5.3.24]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.24.jar:5.3.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:891) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1784) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.69.jar:9.0.69]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: com.fasterxml.jackson.databind.JsonMappingException: The class with com.seeyouletter.api_member.auth.config.PrincipalDetails and name of com.seeyouletter.api_member.auth.config.PrincipalDetails is not in the allowlist. If you believe this class is safe to deserialize, please provide an explicit mapping using Jackson annotations or by providing a Mixin. If the serialization is only done by a trusted source, you can also enable default typing. See https://github.com/spring-projects/spring-security/issues/4370 for details (through reference chain: com.seeyouletter.api_member.config.RedisOauth2AuthorizationService$Authorization["attributes"])
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:392) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:351) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1821) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:316) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:215) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:187) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:144) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:110) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeWithType(BeanDeserializerBase.java:1292) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4674) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3629) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3597) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.seeyouletter.api_member.config.RedisOauth2AuthorizationService.deserialize(RedisOauth2AuthorizationService.java:492) ~[classes/:na]
	... 79 common frames omitted
Caused by: java.lang.IllegalArgumentException: The class with com.seeyouletter.api_member.auth.config.PrincipalDetails and name of com.seeyouletter.api_member.auth.config.PrincipalDetails is not in the allowlist. If you believe this class is safe to deserialize, please provide an explicit mapping using Jackson annotations or by providing a Mixin. If the serialization is only done by a trusted source, you can also enable default typing. See https://github.com/spring-projects/spring-security/issues/4370 for details
	at org.springframework.security.jackson2.SecurityJackson2Modules$AllowlistTypeIdResolver.typeFromId(SecurityJackson2Modules.java:265) ~[spring-security-core-5.7.5.jar:5.7.5]
	at com.fasterxml.jackson.databind.jsontype.impl.TypeDeserializerBase._findDeserializer(TypeDeserializerBase.java:159) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:125) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:110) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromAny(AsPropertyTypeDeserializer.java:213) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserializeWithType(UntypedObjectDeserializer.java:781) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4650) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2831) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at org.springframework.security.jackson2.UsernamePasswordAuthenticationTokenDeserializer.getPrincipal(UsernamePasswordAuthenticationTokenDeserializer.java:104) ~[spring-security-core-5.7.5.jar:5.7.5]
	at org.springframework.security.jackson2.UsernamePasswordAuthenticationTokenDeserializer.deserialize(UsernamePasswordAuthenticationTokenDeserializer.java:75) ~[spring-security-core-5.7.5.jar:5.7.5]
	at org.springframework.security.jackson2.UsernamePasswordAuthenticationTokenDeserializer.deserialize(UsernamePasswordAuthenticationTokenDeserializer.java:51) ~[spring-security-core-5.7.5.jar:5.7.5]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:144) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:110) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromAny(AsPropertyTypeDeserializer.java:213) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.std.UntypedObjectDeserializer$Vanilla.deserializeWithType(UntypedObjectDeserializer.java:781) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.impl.TypeWrappedDeserializer.deserialize(TypeWrappedDeserializer.java:74) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:323) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:4650) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2831) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at org.springframework.security.oauth2.server.authorization.jackson2.UnmodifiableMapDeserializer.deserialize(UnmodifiableMapDeserializer.java:47) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at org.springframework.security.oauth2.server.authorization.jackson2.UnmodifiableMapDeserializer.deserialize(UnmodifiableMapDeserializer.java:37) ~[spring-security-oauth2-authorization-server-0.4.0.jar:0.4.0]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer._deserializeTypedForId(AsPropertyTypeDeserializer.java:144) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.jsontype.impl.AsPropertyTypeDeserializer.deserializeTypedFromObject(AsPropertyTypeDeserializer.java:110) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserializeWithType(MapDeserializer.java:482) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:147) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:314) ~[jackson-databind-2.13.4.2.jar:2.13.4.2]
	... 90 common frames omitted
```

### 버그 설명

엑세스 토큰 발급을 하는 과정에서 토큰을 Redis에 저장하는 `RedisOauth2AuthorizationService`는 JSON 직렬화를 수행한 후에 Redis에 저장하고 있습니다. 로그인된 유저의 정보를 가지고 있는 인증 객체인 `PrincipalDetails`는 Oauth2 인가를 수행하는 과정에서 `RedisOauth2AuthorizationService`에 의해 `ObjectMapper`가 직렬화를 수행하고 Redis에 저장하게 되는데 엑세스 토큰을 생성하기 위해 Redis에 저장되는 정보를 읽어오고 애플리케이션에서 역직렬화를 수행하는 과정에서 `ObejctMapper`가 역직렬화를 수행할 수 있는 대상이 아니므로 에러가 발생하게 됩니다. 
Oauth2 인가를 수행하는 과정에서 직렬화/역직렬화를 수행해야하는 대상 객체는 `PrincipalDetails` 객체를 포함하고 있는데, 스프링 시큐리티에서 사용하는 인가 및 인증 관련 클래스(3rd party 클래스)들은 `ObjectMapper`가 역직렬화를 수행할 수 있도록 기본 생성자를 가진 형태가 아니기 때문에 스프링 시큐리티에서는 대상 객체들을 역직렬화할 수 있도록 `ObjectMapper`가 사용할 수 있는 모듈([SecurityJackson2Modules](https://github.com/spring-projects/spring-security/blob/main/core/src/main/java/org/springframework/security/jackson2/SecurityJackson2Modules.java), [OAuth2AuthorizationServerJackson2Module](https://github.com/spring-projects/spring-authorization-server/blob/main/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/jackson2/OAuth2AuthorizationServerJackson2Module.java))들을 제공하고 있습니다. 해당 모듈은 `mixIn`을 통해서 객체를 역직렬화할 수 있도록 지원하지만 인증 관련 객체가 대상이기 때문에 [허용된 객체(mixIn)들을 대상으로만 역직렬화를 수행할 수 있도록 구현](https://github.com/spring-projects/spring-security/blob/be2958ed13698fa65aa552f3df71735c0107eed4/core/src/main/java/org/springframework/security/jackson2/SecurityJackson2Modules.java#L276-L289)되어 있습니다.

다만 테스트 코드가 실패하지 않은 이유는 테스트 코드에서는 유저의 인증을 mock하기 위해 `@WithMockUser` 애노테이션을 사용하고 있습니다. 해당 애노테이션은 [WithMockUserSecurityContextFactory](https://github.com/spring-projects/spring-security/blob/main/test/src/main/java/org/springframework/security/test/context/support/WithMockUserSecurityContextFactory.java)을 통해서 mock 유저 인증 객체를 생성하고 `SecurityContextHolder`에 주입합니다. 하지만 현재 애플리케이션에서 인증된 유저 객체는 `PrincipalDetails`인데 테스트 환경에서는 이 객체가 사용되지 않았기 때문에 통과할 수 있었습니다.

### 예상 결과

`PrincipalDetails` 클래스를 `RedisOauth2AuthorizationService`에서 사용하는 `ObjectMapper`가 역직렬화할 수 있도록 `mixIn`으로 추가하여 정상적으로 역직렬화를 수행할 수 있도록 해결할 수 있습니다. 
테스트 환경에서는 mock 인증 객체를 `PrincipalDetails` 객체를 사용하도록 커스텀한 `WithSecurityContextFactory` 구현체와 애노테이션을 적용하여 실제 환경과 테스트 환경에서 동일한 코드로 테스트하여 정확한 테스트를 수행할 수 있도록 보완할 수 있습니다.

### 디바이스 환경

환경에 구분없이 발생하는 에러입니다.

### 🌱 브랜치 입력(개발 작업 중인 경우)

`fix/redis-authorization-service-custom-principal-json-deserialize`

--- a/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomAuthorization.java
+++ b/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomAuthorization.java
@@ -1,0 +1,75 @@
+package com.seeyouletter.api_member.auth.config;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.Id;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Set;
+
+import static lombok.AccessLevel.PRIVATE;
+import static lombok.AccessLevel.PROTECTED;
+
+@Getter
+@Builder
+@NoArgsConstructor(access = PROTECTED)
+@AllArgsConstructor(access = PRIVATE)
+public class CustomAuthorization {
+
+    @Id
+    private String id;
+
+    private String registeredClientId;
+
+    private String principalName;
+
+    private String authorizationGrantType;
+
+    private Set<String> authorizedScopes;
+
+    private Map<String, Object> attributes;
+
+    private String state;
+
+    private String authorizationCodeValue;
+
+    private Instant authorizationCodeIssuedAt;
+
+    private Instant authorizationCodeExpiresAt;
+
+    private Map<String, Object> authorizationCodeMetadata;
+
+    private String accessTokenValue;
+
+    private Instant accessTokenIssuedAt;
+
+    private Instant accessTokenExpiresAt;
+
+    private Map<String, Object> accessTokenMetadata;
+
+    private String accessTokenType;
+
+    private Set<String> accessTokenScopes;
+
+    private String refreshTokenValue;
+
+    private Instant refreshTokenIssuedAt;
+
+    private Instant refreshTokenExpiresAt;
+
+    private Map<String, Object> refreshTokenMetadata;
+
+    private String oidcIdTokenValue;
+
+    private Instant oidcIdTokenIssuedAt;
+
+    private Instant oidcIdTokenExpiresAt;
+
+    private Map<String, Object> oidcIdTokenMetadata;
+
+    private Map<String, Object> oidcIdTokenClaims;
+
+}

--- a/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomAuthorizationMixIn.java
+++ b/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomAuthorizationMixIn.java
@@ -1,0 +1,25 @@
+package com.seeyouletter.api_member.auth.config;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+import static com.fasterxml.jackson.annotation.JsonTypeInfo.Id.CLASS;
+
+@JsonTypeInfo(use = CLASS)
+@JsonAutoDetect(
+        fieldVisibility = ANY,
+        getterVisibility = NONE,
+        isGetterVisibility = NONE
+)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public abstract class CustomAuthorizationMixIn {
+
+    @JsonCreator
+    CustomAuthorizationMixIn() {
+    }
+
+}

--- a/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomOauth2UserMixIn.java
+++ b/api-member/src/main/java/com/seeyouletter/api_member/auth/config/CustomOauth2UserMixIn.java
@@ -1,0 +1,28 @@
+package com.seeyouletter.api_member.auth.config;
+
+import com.fasterxml.jackson.annotation.*;
+import org.springframework.security.core.GrantedAuthority;
+
+import java.util.Collection;
+import java.util.Map;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+import static com.fasterxml.jackson.annotation.JsonTypeInfo.Id.CLASS;
+
+@JsonTypeInfo(use = CLASS)
+@JsonAutoDetect(
+        fieldVisibility = ANY,
+        getterVisibility = NONE,
+        isGetterVisibility = NONE
+)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public abstract class CustomOauth2UserMixIn {
+
+    @JsonCreator
+    CustomOauth2UserMixIn(@JsonProperty(value = "authorities") Collection<? extends GrantedAuthority> authorities,
+                          @JsonProperty(value = "attributes") Map<String, Object> attributes,
+                          @JsonProperty(value = "username") String username) {
+    }
+
+}

--- a/api-member/src/main/java/com/seeyouletter/api_member/config/RedisOauth2AuthorizationService.java
+++ b/api-member/src/main/java/com/seeyouletter/api_member/config/RedisOauth2AuthorizationService.java
@@ -1,17 +1,13 @@
 package com.seeyouletter.api_member.config;
 
-import com.fasterxml.jackson.annotation.JsonAutoDetect;
-import com.fasterxml.jackson.annotation.JsonCreator;
-import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
-import com.fasterxml.jackson.annotation.JsonTypeInfo;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
+import com.seeyouletter.api_member.auth.config.CustomAuthorization;
+import com.seeyouletter.api_member.auth.config.CustomAuthorization.CustomAuthorizationBuilder;
+import com.seeyouletter.api_member.auth.config.CustomAuthorizationMixIn;
+import com.seeyouletter.api_member.auth.config.CustomOAuth2User;
+import com.seeyouletter.api_member.auth.config.CustomOauth2UserMixIn;
 import org.springframework.dao.DataRetrievalFailureException;
-import org.springframework.data.annotation.Id;
 import org.springframework.data.redis.connection.RedisStringCommands;
 import org.springframework.data.redis.core.RedisCallback;
 import org.springframework.data.redis.core.StringRedisTemplate;
@@ -33,19 +29,11 @@ import org.springframework.util.Assert;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
 import java.time.Duration;
-import java.time.Instant;
-import java.util.Map;
-import java.util.Set;
 
-import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
-import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
-import static com.fasterxml.jackson.annotation.JsonTypeInfo.Id.CLASS;
 import static java.lang.String.format;
 import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.security.MessageDigest.getInstance;
 import static java.time.temporal.ChronoUnit.MINUTES;
-import static lombok.AccessLevel.PRIVATE;
-import static lombok.AccessLevel.PROTECTED;
 import static org.springframework.data.redis.connection.RedisStringCommands.SetOption.UPSERT;
 import static org.springframework.data.redis.core.types.Expiration.from;
 import static org.springframework.security.jackson2.SecurityJackson2Modules.getModules;
@@ -78,7 +66,8 @@ public final class RedisOauth2AuthorizationService implements OAuth2Authorizatio
     private final ObjectMapper objectMapper = new ObjectMapper()
             .registerModules(getModules(this.getClass().getClassLoader()))
             .registerModule(new OAuth2AuthorizationServerJackson2Module())
-            .addMixIn(Authorization.class, AuthorizationMixIn.class);
+            .addMixIn(CustomAuthorization.class, CustomAuthorizationMixIn.class)
+            .addMixIn(CustomOAuth2User.class, CustomOauth2UserMixIn.class);
 
     private final RegisteredClientRepository registeredClientRepository;
 
@@ -336,7 +325,7 @@ public final class RedisOauth2AuthorizationService implements OAuth2Authorizatio
         return getClientRefreshTokenTimeToLive(registeredClient);
     }
 
-    private OAuth2Authorization toObject(Authorization entity) {
+    private OAuth2Authorization toObject(CustomAuthorization entity) {
         OAuth2Authorization.Builder builder = OAuth2Authorization
                 .withRegisteredClient(findClientByRegisteredClientId(entity.getRegisteredClientId()))
                 .id(entity.getId())
@@ -395,8 +384,8 @@ public final class RedisOauth2AuthorizationService implements OAuth2Authorizatio
         return builder.build();
     }
 
-    private Authorization toEntity(OAuth2Authorization authorization) {
-        Authorization.AuthorizationBuilder builder = Authorization
+    private CustomAuthorization toEntity(OAuth2Authorization authorization) {
+        CustomAuthorizationBuilder builder = CustomAuthorization
                 .builder()
                 .id(authorization.getId())
                 .registeredClientId(authorization.getRegisteredClientId())
@@ -489,7 +478,7 @@ public final class RedisOauth2AuthorizationService implements OAuth2Authorizatio
         }
 
         try {
-            return toObject(objectMapper.readValue(json, Authorization.class));
+            return toObject(objectMapper.readValue(json, CustomAuthorization.class));
         } catch (JsonProcessingException e) {
             throw new IllegalArgumentException(e.getMessage(), e);
         }
@@ -517,82 +506,6 @@ public final class RedisOauth2AuthorizationService implements OAuth2Authorizatio
         }
 
         return stringBuilder.toString();
-    }
-
-    @Getter
-    @Builder
-    @NoArgsConstructor(access = PROTECTED)
-    @AllArgsConstructor(access = PRIVATE)
-    public static class Authorization {
-
-        @Id
-        private String id;
-
-        private String registeredClientId;
-
-        private String principalName;
-
-        private String authorizationGrantType;
-
-        private Set<String> authorizedScopes;
-
-        private Map<String, Object> attributes;
-
-        private String state;
-
-        private String authorizationCodeValue;
-
-        private Instant authorizationCodeIssuedAt;
-
-        private Instant authorizationCodeExpiresAt;
-
-        private Map<String, Object> authorizationCodeMetadata;
-
-        private String accessTokenValue;
-
-        private Instant accessTokenIssuedAt;
-
-        private Instant accessTokenExpiresAt;
-
-        private Map<String, Object> accessTokenMetadata;
-
-        private String accessTokenType;
-
-        private Set<String> accessTokenScopes;
-
-        private String refreshTokenValue;
-
-        private Instant refreshTokenIssuedAt;
-
-        private Instant refreshTokenExpiresAt;
-
-        private Map<String, Object> refreshTokenMetadata;
-
-        private String oidcIdTokenValue;
-
-        private Instant oidcIdTokenIssuedAt;
-
-        private Instant oidcIdTokenExpiresAt;
-
-        private Map<String, Object> oidcIdTokenMetadata;
-
-        private Map<String, Object> oidcIdTokenClaims;
-
-    }
-
-    @JsonTypeInfo(use = CLASS)
-    @JsonAutoDetect(
-            fieldVisibility = ANY,
-            getterVisibility = NONE,
-            isGetterVisibility = NONE
-    )
-    @JsonIgnoreProperties(ignoreUnknown = true)
-    public static abstract class AuthorizationMixIn {
-
-        @JsonCreator
-        AuthorizationMixIn() {
-        }
-
     }
 
 }

--- a/api-member/src/test/java/com/seeyouletter/api_member/config/WithMockOauth2User.java
+++ b/api-member/src/test/java/com/seeyouletter/api_member/config/WithMockOauth2User.java
@@ -1,0 +1,47 @@
+package com.seeyouletter.api_member.config;
+
+import org.springframework.core.annotation.AliasFor;
+import org.springframework.security.test.context.support.TestExecutionEvent;
+import org.springframework.security.test.context.support.WithSecurityContext;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static org.springframework.security.test.context.support.TestExecutionEvent.TEST_METHOD;
+
+@Target(value = {METHOD, TYPE})
+@Retention(value = RUNTIME)
+@Inherited
+@Documented
+@WithSecurityContext(factory = WithMockOauth2UserSecurityContextFactory.class)
+public @interface WithMockOauth2User {
+
+    String value() default "user";
+
+    String username() default "";
+
+    String[] roles() default {"USER"};
+
+    String[] authorities() default {};
+
+    String clientRegistrationId() default "client";
+
+    AttributeKeyPair[] attributes() default {};
+
+    @AliasFor(annotation = WithSecurityContext.class)
+    TestExecutionEvent setupBefore() default TEST_METHOD;
+
+    @interface AttributeKeyPair {
+
+        String key();
+
+        String value();
+
+    }
+
+}

--- a/api-member/src/test/java/com/seeyouletter/api_member/config/WithMockOauth2UserSecurityContextFactory.java
+++ b/api-member/src/test/java/com/seeyouletter/api_member/config/WithMockOauth2UserSecurityContextFactory.java
@@ -1,0 +1,91 @@
+package com.seeyouletter.api_member.config;
+
+import com.seeyouletter.api_member.auth.config.CustomOAuth2User;
+import com.seeyouletter.api_member.config.WithMockOauth2User.AttributeKeyPair;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
+import org.springframework.security.test.context.support.WithSecurityContextFactory;
+import org.springframework.util.Assert;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+import static org.springframework.util.StringUtils.hasLength;
+
+public class WithMockOauth2UserSecurityContextFactory implements WithSecurityContextFactory<WithMockOauth2User> {
+
+    @Override
+    public SecurityContext createSecurityContext(WithMockOauth2User withUser) {
+        SecurityContext context = SecurityContextHolder.createEmptyContext();
+
+        context.setAuthentication(createAuthentication(withUser));
+
+        return context;
+    }
+
+    private String getUsername(WithMockOauth2User withUser) {
+        if (hasLength(withUser.username())) {
+            return withUser.username();
+        }
+
+        Assert.notNull(withUser.value(), () -> withUser + " cannot have null username on both username and value properties");
+
+        return withUser.value();
+    }
+
+    private List<GrantedAuthority> createGrantedAuthorities(WithMockOauth2User withUser) {
+        List<GrantedAuthority> grantedAuthorities = Arrays
+                .stream(withUser.authorities())
+                .map(SimpleGrantedAuthority::new)
+                .collect(toList());
+
+        if (grantedAuthorities.isEmpty()) {
+            for (String role : withUser.roles()) {
+                Assert.isTrue(!role.startsWith("ROLE_"), () -> "roles cannot start with ROLE_ Got " + role);
+
+                grantedAuthorities.add(new SimpleGrantedAuthority("ROLE_" + role));
+            }
+        }
+
+        return grantedAuthorities;
+    }
+
+    private Map<String, Object> createAttributes(WithMockOauth2User withUser) {
+        Map<String, Object> attributes = Arrays
+                .stream(withUser.attributes())
+                .collect(
+                        Collectors.toMap(
+                                AttributeKeyPair::key,
+                                AttributeKeyPair::value
+                        )
+                );
+
+        if (attributes.isEmpty()) {
+            attributes.put("id", getUsername(withUser));
+        }
+
+        return attributes;
+    }
+
+    private Authentication createAuthentication(WithMockOauth2User withUser) {
+        if (!(withUser.roles().length == 1 && "USER".equals(withUser.roles()[0]))) {
+            throw new IllegalStateException("You cannot define roles attribute " + asList(withUser.roles())
+                    + " with authorities attribute " + asList(withUser.authorities()));
+        }
+
+        List<GrantedAuthority> grantedAuthorities = createGrantedAuthorities(withUser);
+        CustomOAuth2User customOAuth2User = new CustomOAuth2User(grantedAuthorities, createAttributes(withUser), getUsername(withUser));
+
+        return new OAuth2AuthenticationToken(customOAuth2User, grantedAuthorities, withUser.clientRegistrationId());
+    }
+
+
+}

--- a/api-member/src/test/java/com/seeyouletter/api_member/e2e/Oauth2AuthorizationTest.java
+++ b/api-member/src/test/java/com/seeyouletter/api_member/e2e/Oauth2AuthorizationTest.java
@@ -2,12 +2,10 @@ package com.seeyouletter.api_member.e2e;
 
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.seeyouletter.api_member.IntegrationTestContext;
+import com.seeyouletter.api_member.config.WithMockOauth2User;
 import org.apache.http.NameValuePair;
 import org.apache.http.client.utils.URLEncodedUtils;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.data.redis.core.StringRedisTemplate;
 import org.springframework.security.oauth2.jwt.Jwt;
@@ -79,290 +77,466 @@ class Oauth2AuthorizationTest extends IntegrationTestContext {
                 .flushAll();
     }
 
-    @Test
+    @Nested
     @WithMockUser
-    @DisplayName(value = "authorization")
-    void authorization() throws Exception {
-        // given
-        String clientId = publicClient.getClientId();
-        String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
-        String scope = join(" ", publicClient.getScopes());
-        String codeVerifier = generateCodeVerifier();
-        String codeChallenge = generateCodeChallenge(codeVerifier);
-        String state = randomUUID().toString();
-        String nonce = randomUUID().toString();
+    @DisplayName(value = "first party 유저 세션의 Oauth2 인증 및 인가")
+    class FirstPartyUserSessionOauth2Authorization {
 
-        // when & then
-        MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
-                .andDo(
-                        document(
-                                "authorization",
-                                REQUEST_PREPROCESSOR,
-                                RESPONSE_PREPROCESSOR,
-                                requestParameters(
-                                        parameterWithName("client_id").description("클라이언트 id"),
-                                        parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
-                                        parameterWithName("scope").description("토큰의 인가 범위").optional(),
-                                        parameterWithName("response_type").description("응답 유형, code 고정으로 사용"),
-                                        parameterWithName("code_challenge").description("해시 값, Base64(SHA256(code_verifier))"),
-                                        parameterWithName("code_challenge_method").description("해시 방식, S256 고정으로 사용"),
-                                        parameterWithName("state").description("리다이렉트 callback uri로 전달되는 값").optional(),
-                                        parameterWithName("nonce").description("id_token claim에 포함되는 값").optional()
-                                ),
-                                responseHeaders(
-                                        headerWithName(LOCATION).description(LOCATION)
-                                )
-                        )
-                )
-                .andReturn();
+        @Test
+        @DisplayName(value = "authorization")
+        void authorization() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
 
-        Map<String, String> queryStrings = parseRedirectQueryString(authorizationResult);
+            // when & then
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andDo(
+                            document(
+                                    "authorization",
+                                    REQUEST_PREPROCESSOR,
+                                    RESPONSE_PREPROCESSOR,
+                                    requestParameters(
+                                            parameterWithName("client_id").description("클라이언트 id"),
+                                            parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
+                                            parameterWithName("scope").description("토큰의 인가 범위").optional(),
+                                            parameterWithName("response_type").description("응답 유형, code 고정으로 사용"),
+                                            parameterWithName("code_challenge").description("해시 값, Base64(SHA256(code_verifier))"),
+                                            parameterWithName("code_challenge_method").description("해시 방식, S256 고정으로 사용"),
+                                            parameterWithName("state").description("리다이렉트 callback uri로 전달되는 값").optional(),
+                                            parameterWithName("nonce").description("id_token claim에 포함되는 값").optional()
+                                    ),
+                                    responseHeaders(
+                                            headerWithName(LOCATION).description(LOCATION)
+                                    )
+                            )
+                    )
+                    .andReturn();
 
-        assertThat(queryStrings.get("state")).isEqualTo(state);
-        assertThat(queryStrings.get("code")).isNotEmpty();
+            Map<String, String> queryStrings = parseRedirectQueryString(authorizationResult);
+
+            assertThat(queryStrings.get("state")).isEqualTo(state);
+            assertThat(queryStrings.get("code")).isNotEmpty();
+        }
+
+        @Test
+        @DisplayName(value = "token")
+        void token() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
+
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
+
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
+
+            // when & then
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andDo(
+                            document(
+                                    "token",
+                                    REQUEST_PREPROCESSOR,
+                                    RESPONSE_PREPROCESSOR,
+                                    requestHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    requestParameters(
+                                            parameterWithName("client_id").description("클라이언트 id"),
+                                            parameterWithName("code").description("인가 코드"),
+                                            parameterWithName("code_verifier").description("해시 원본 값"),
+                                            parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
+                                            parameterWithName("redirect_uri").description("리다이렉트 callback uri")
+                                    ),
+                                    responseHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    responseFields(
+                                            fieldWithPath("access_token").description("엑세스 토큰"),
+                                            fieldWithPath("scope").description("엑세스 토큰의 인가 범위"),
+                                            fieldWithPath("id_token").description("인증 토큰, 인가 요청시 scope로 openid를 전달한 경우에만 발급").optional(),
+                                            fieldWithPath("token_type").description("토큰 타입, Bearer 고정으로 사용"),
+                                            fieldWithPath("expires_in").description("토큰의 남은 유효기간")
+                                    )
+                            )
+                    )
+                    .andReturn();
+
+            Map<String, Object> fields = parsePayloadFields(tokenResult);
+            Jwt idToken = jwtDecoder.decode((String) fields.get("id_token"));
+
+            assertThat(nonce).isEqualTo(idToken.getClaim("nonce"));
+        }
+
+        @Test
+        @DisplayName(value = "introspect")
+        void introspect() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
+
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
+
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
+
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
+
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
+
+            // when & then
+            mockMvc
+                    .perform(
+                            post("/oauth2/introspect")
+                                    .contentType(APPLICATION_FORM_URLENCODED)
+                                    .param("client_id", clientId)
+                                    .param("code", code)
+                                    .param("code_verifier", codeVerifier)
+                                    .param("grant_type", "authorization_code")
+                                    .param("redirect_uri", redirectUri)
+                                    .param("token", accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print())
+                    .andDo(
+                            document(
+                                    "introspect",
+                                    REQUEST_PREPROCESSOR,
+                                    RESPONSE_PREPROCESSOR,
+                                    requestHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    requestParameters(
+                                            parameterWithName("client_id").description("클라이언트 id"),
+                                            parameterWithName("code").description("인가 코드"),
+                                            parameterWithName("code_verifier").description("해시 원본 값"),
+                                            parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
+                                            parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
+                                            parameterWithName("token").description("엑세스 토큰")
+                                    ),
+                                    responseHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    responseFields(
+                                            fieldWithPath("active").description("엑세스 토큰의 유효 여부"),
+                                            fieldWithPath("sub").description("인가 요청자"),
+                                            fieldWithPath("aud").description("인가 클라이언트"),
+                                            fieldWithPath("nbf").description("엑세스 토큰이 활성화된 시간(unix time)"),
+                                            fieldWithPath("scope").description("엑세스 토큰의 인가 범위"),
+                                            fieldWithPath("iss").description("엑세스 토큰 발행자"),
+                                            fieldWithPath("exp").description("엑세스 토큰이 만료되는 시간(unix time)"),
+                                            fieldWithPath("iat").description("엑세스 토큰이 발행된 시간(unix time)"),
+                                            fieldWithPath("client_id").description("클라이언트 id"),
+                                            fieldWithPath("token_type").description("토큰 타입, Bearer 고정으로 사용")
+                                    )
+                            )
+                    );
+        }
+
+        @Test
+        @DisplayName(value = "revoke")
+        void revoke() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
+
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
+
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
+
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
+
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
+
+            // when & then
+            mockMvc
+                    .perform(
+                            post("/oauth2/revoke")
+                                    .contentType(APPLICATION_FORM_URLENCODED)
+                                    .param("client_id", clientId)
+                                    .param("code", code)
+                                    .param("code_verifier", codeVerifier)
+                                    .param("grant_type", "authorization_code")
+                                    .param("redirect_uri", redirectUri)
+                                    .param("token", accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print())
+                    .andDo(
+                            document(
+                                    "revoke",
+                                    REQUEST_PREPROCESSOR,
+                                    RESPONSE_PREPROCESSOR,
+                                    requestHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    requestParameters(
+                                            parameterWithName("client_id").description("클라이언트 id"),
+                                            parameterWithName("code").description("인가 코드"),
+                                            parameterWithName("code_verifier").description("해시 원본 값"),
+                                            parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
+                                            parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
+                                            parameterWithName("token").description("엑세스 토큰")
+                                    )
+                            )
+                    );
+        }
+
+        @Test
+        @DisplayName(value = "userinfo")
+        void userinfo() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
+
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
+
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
+
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
+
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
+
+            // when & then
+            mockMvc
+                    .perform(
+                            get("/userinfo")
+                                    .header(AUTHORIZATION, "Bearer " + accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print())
+                    .andDo(
+                            document(
+                                    "userinfo",
+                                    REQUEST_PREPROCESSOR,
+                                    RESPONSE_PREPROCESSOR,
+                                    requestHeaders(
+                                            headerWithName(AUTHORIZATION).description("Bearer {access_token}")
+                                    ),
+                                    responseHeaders(
+                                            headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
+                                    ),
+                                    responseFields(
+                                            fieldWithPath("sub").description("인가 요청자")
+                                    )
+                            )
+                    );
+        }
+
     }
 
-    @Test
-    @WithMockUser
-    @DisplayName(value = "token")
-    void token() throws Exception {
-        // given
-        String clientId = publicClient.getClientId();
-        String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
-        String scope = join(" ", publicClient.getScopes());
-        String codeVerifier = generateCodeVerifier();
-        String codeChallenge = generateCodeChallenge(codeVerifier);
-        String state = randomUUID().toString();
-        String nonce = randomUUID().toString();
+    @Nested
+    @WithMockOauth2User
+    @DisplayName(value = "third party 유저 세션의 Oauth2 인증 및 인가")
+    class ThirdPartyUserSessionOauth2Authorization {
 
-        MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
-                .andReturn();
+        @Test
+        @DisplayName(value = "authorization")
+        void authorization() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
 
-        String code = parseRedirectQueryString(authorizationResult)
-                .get("code");
+            // when & then
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
 
-        // when & then
-        MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
-                .andDo(
-                        document(
-                                "token",
-                                REQUEST_PREPROCESSOR,
-                                RESPONSE_PREPROCESSOR,
-                                requestHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                requestParameters(
-                                        parameterWithName("client_id").description("클라이언트 id"),
-                                        parameterWithName("code").description("인가 코드"),
-                                        parameterWithName("code_verifier").description("해시 원본 값"),
-                                        parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
-                                        parameterWithName("redirect_uri").description("리다이렉트 callback uri")
-                                ),
-                                responseHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                responseFields(
-                                        fieldWithPath("access_token").description("엑세스 토큰"),
-                                        fieldWithPath("scope").description("엑세스 토큰의 인가 범위"),
-                                        fieldWithPath("id_token").description("인증 토큰, 인가 요청시 scope로 openid를 전달한 경우에만 발급").optional(),
-                                        fieldWithPath("token_type").description("토큰 타입, Bearer 고정으로 사용"),
-                                        fieldWithPath("expires_in").description("토큰의 남은 유효기간")
-                                )
-                        )
-                )
-                .andReturn();
+            Map<String, String> queryStrings = parseRedirectQueryString(authorizationResult);
 
-        Map<String, Object> fields = parsePayloadFields(tokenResult);
-        Jwt idToken = jwtDecoder.decode((String) fields.get("id_token"));
+            assertThat(queryStrings.get("state")).isEqualTo(state);
+            assertThat(queryStrings.get("code")).isNotEmpty();
+        }
 
-        assertThat(nonce).isEqualTo(idToken.getClaim("nonce"));
-    }
+        @Test
+        @DisplayName(value = "token")
+        void token() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
 
-    @Test
-    @WithMockUser
-    @DisplayName(value = "introspect")
-    void introspect() throws Exception {
-        // given
-        String clientId = publicClient.getClientId();
-        String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
-        String scope = join(" ", publicClient.getScopes());
-        String codeVerifier = generateCodeVerifier();
-        String codeChallenge = generateCodeChallenge(codeVerifier);
-        String state = randomUUID().toString();
-        String nonce = randomUUID().toString();
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
 
-        MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
-                .andReturn();
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
 
-        String code = parseRedirectQueryString(authorizationResult)
-                .get("code");
+            // when & then
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
 
-        MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
-                .andReturn();
+            Map<String, Object> fields = parsePayloadFields(tokenResult);
+            Jwt idToken = jwtDecoder.decode((String) fields.get("id_token"));
 
-        String accessTokenValue = jwtDecoder
-                .decode((String) parsePayloadFields(tokenResult).get("access_token"))
-                .getTokenValue();
+            assertThat(nonce).isEqualTo(idToken.getClaim("nonce"));
+        }
 
-        // when & then
-        mockMvc
-                .perform(
-                        post("/oauth2/introspect")
-                                .contentType(APPLICATION_FORM_URLENCODED)
-                                .param("client_id", clientId)
-                                .param("code", code)
-                                .param("code_verifier", codeVerifier)
-                                .param("grant_type", "authorization_code")
-                                .param("redirect_uri", redirectUri)
-                                .param("token", accessTokenValue)
-                )
-                .andExpect(status().isOk())
-                .andDo(print())
-                .andDo(
-                        document(
-                                "introspect",
-                                REQUEST_PREPROCESSOR,
-                                RESPONSE_PREPROCESSOR,
-                                requestHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                requestParameters(
-                                        parameterWithName("client_id").description("클라이언트 id"),
-                                        parameterWithName("code").description("인가 코드"),
-                                        parameterWithName("code_verifier").description("해시 원본 값"),
-                                        parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
-                                        parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
-                                        parameterWithName("token").description("엑세스 토큰")
-                                ),
-                                responseHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                responseFields(
-                                        fieldWithPath("active").description("엑세스 토큰의 유효 여부"),
-                                        fieldWithPath("sub").description("인가 요청자"),
-                                        fieldWithPath("aud").description("인가 클라이언트"),
-                                        fieldWithPath("nbf").description("엑세스 토큰이 활성화된 시간(unix time)"),
-                                        fieldWithPath("scope").description("엑세스 토큰의 인가 범위"),
-                                        fieldWithPath("iss").description("엑세스 토큰 발행자"),
-                                        fieldWithPath("exp").description("엑세스 토큰이 만료되는 시간(unix time)"),
-                                        fieldWithPath("iat").description("엑세스 토큰이 발행된 시간(unix time)"),
-                                        fieldWithPath("client_id").description("클라이언트 id"),
-                                        fieldWithPath("token_type").description("토큰 타입, Bearer 고정으로 사용")
-                                )
-                        )
-                );
-    }
+        @Test
+        @DisplayName(value = "introspect")
+        void introspect() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
 
-    @Test
-    @WithMockUser
-    @DisplayName(value = "revoke")
-    void revoke() throws Exception {
-        // given
-        String clientId = publicClient.getClientId();
-        String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
-        String scope = join(" ", publicClient.getScopes());
-        String codeVerifier = generateCodeVerifier();
-        String codeChallenge = generateCodeChallenge(codeVerifier);
-        String state = randomUUID().toString();
-        String nonce = randomUUID().toString();
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
 
-        MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
-                .andReturn();
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
 
-        String code = parseRedirectQueryString(authorizationResult)
-                .get("code");
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
 
-        MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
-                .andReturn();
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
 
-        String accessTokenValue = jwtDecoder
-                .decode((String) parsePayloadFields(tokenResult).get("access_token"))
-                .getTokenValue();
+            // when & then
+            mockMvc
+                    .perform(
+                            post("/oauth2/introspect")
+                                    .contentType(APPLICATION_FORM_URLENCODED)
+                                    .param("client_id", clientId)
+                                    .param("code", code)
+                                    .param("code_verifier", codeVerifier)
+                                    .param("grant_type", "authorization_code")
+                                    .param("redirect_uri", redirectUri)
+                                    .param("token", accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print());
+        }
 
-        // when & then
-        mockMvc
-                .perform(
-                        post("/oauth2/revoke")
-                                .contentType(APPLICATION_FORM_URLENCODED)
-                                .param("client_id", clientId)
-                                .param("code", code)
-                                .param("code_verifier", codeVerifier)
-                                .param("grant_type", "authorization_code")
-                                .param("redirect_uri", redirectUri)
-                                .param("token", accessTokenValue)
-                )
-                .andExpect(status().isOk())
-                .andDo(print())
-                .andDo(
-                        document(
-                                "revoke",
-                                REQUEST_PREPROCESSOR,
-                                RESPONSE_PREPROCESSOR,
-                                requestHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                requestParameters(
-                                        parameterWithName("client_id").description("클라이언트 id"),
-                                        parameterWithName("code").description("인가 코드"),
-                                        parameterWithName("code_verifier").description("해시 원본 값"),
-                                        parameterWithName("grant_type").description("인증 방식, authorization_code 고정으로 사용"),
-                                        parameterWithName("redirect_uri").description("리다이렉트 callback uri"),
-                                        parameterWithName("token").description("엑세스 토큰")
-                                )
-                        )
-                );
-    }
+        @Test
+        @DisplayName(value = "revoke")
+        void revoke() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
 
-    @Test
-    @WithMockUser
-    @DisplayName(value = "userinfo")
-    void userinfo() throws Exception {
-        // given
-        String clientId = publicClient.getClientId();
-        String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
-        String scope = join(" ", publicClient.getScopes());
-        String codeVerifier = generateCodeVerifier();
-        String codeChallenge = generateCodeChallenge(codeVerifier);
-        String state = randomUUID().toString();
-        String nonce = randomUUID().toString();
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
 
-        MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
-                .andReturn();
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
 
-        String code = parseRedirectQueryString(authorizationResult)
-                .get("code");
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
 
-        MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
-                .andReturn();
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
 
-        String accessTokenValue = jwtDecoder
-                .decode((String) parsePayloadFields(tokenResult).get("access_token"))
-                .getTokenValue();
+            // when & then
+            mockMvc
+                    .perform(
+                            post("/oauth2/revoke")
+                                    .contentType(APPLICATION_FORM_URLENCODED)
+                                    .param("client_id", clientId)
+                                    .param("code", code)
+                                    .param("code_verifier", codeVerifier)
+                                    .param("grant_type", "authorization_code")
+                                    .param("redirect_uri", redirectUri)
+                                    .param("token", accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print());
+        }
 
-        // when & then
-        mockMvc
-                .perform(
-                        get("/userinfo")
-                                .header(AUTHORIZATION, "Bearer " + accessTokenValue)
-                )
-                .andExpect(status().isOk())
-                .andDo(print())
-                .andDo(
-                        document(
-                                "userinfo",
-                                REQUEST_PREPROCESSOR,
-                                RESPONSE_PREPROCESSOR,
-                                requestHeaders(
-                                        headerWithName(AUTHORIZATION).description("Bearer {access_token}")
-                                ),
-                                responseHeaders(
-                                        headerWithName(CONTENT_TYPE).description(CONTENT_TYPE)
-                                ),
-                                responseFields(
-                                        fieldWithPath("sub").description("인가 요청자")
-                                )
-                        )
-                );
+        @Test
+        @DisplayName(value = "userinfo")
+        void userinfo() throws Exception {
+            // given
+            String clientId = publicClient.getClientId();
+            String redirectUri = publicClient.getRedirectUris().stream().findFirst().orElseThrow();
+            String scope = join(" ", publicClient.getScopes());
+            String codeVerifier = generateCodeVerifier();
+            String codeChallenge = generateCodeChallenge(codeVerifier);
+            String state = randomUUID().toString();
+            String nonce = randomUUID().toString();
+
+            MvcResult authorizationResult = performAuthorizationRequest(clientId, redirectUri, scope, codeChallenge, state, nonce)
+                    .andReturn();
+
+            String code = parseRedirectQueryString(authorizationResult)
+                    .get("code");
+
+            MvcResult tokenResult = performTokenRequest(clientId, redirectUri, codeVerifier, code)
+                    .andReturn();
+
+            String accessTokenValue = jwtDecoder
+                    .decode((String) parsePayloadFields(tokenResult).get("access_token"))
+                    .getTokenValue();
+
+            // when & then
+            mockMvc
+                    .perform(
+                            get("/userinfo")
+                                    .header(AUTHORIZATION, "Bearer " + accessTokenValue)
+                    )
+                    .andExpect(status().isOk())
+                    .andDo(print());
+        }
+
     }
 
     static RegisteredClient createOauth2PublicClient() {
@@ -395,17 +569,17 @@ class Oauth2AuthorizationTest extends IntegrationTestContext {
                                                       String state,
                                                       String nonce) throws Exception {
         return mockMvc.perform(
-                        get("/oauth2/authorize")
-                                .with(csrf())
-                                .queryParam("client_id", clientId)
-                                .queryParam("redirect_uri", redirectUri)
-                                .queryParam("scope", scope)
-                                .queryParam("response_type", "code")
-                                .queryParam("code_challenge", codeChallenge)
-                                .queryParam("code_challenge_method", "S256")
-                                .queryParam("state", state)
-                                .queryParam("nonce", nonce)
-                )
+                get("/oauth2/authorize")
+                        .with(csrf())
+                        .queryParam("client_id", clientId)
+                        .queryParam("redirect_uri", redirectUri)
+                        .queryParam("scope", scope)
+                        .queryParam("response_type", "code")
+                        .queryParam("code_challenge", codeChallenge)
+                        .queryParam("code_challenge_method", "S256")
+                        .queryParam("state", state)
+                        .queryParam("nonce", nonce)
+        )
                 .andExpect(status().is3xxRedirection())
                 .andDo(print());
     }

